src/hal/src/acceleration_structure.rs

```diff
@@ -0,0 +1,232 @@
+    pub vertex_buffer_offset: Offset,
+    pub vertex_buffer_stride: Offset,
+
+    /// aka "vertex count"?
+    pub max_vertex: Offset,
+
+    // index format must be DXGI_FORMAT_R32_UINT, DXGI_FORMAT_R16_UINT
+    // can also be VK_INDEX_TYPE_NONE_KHR/DXGI_FORMAT_UNKNOWN
+    pub index_buffer: Option<(&'a B::Buffer, Offset, IndexType)>,
+
+    /// 3x4 matrix
+    /// TODO enum for cpu repr?
+    /// Must point to an address containing `TransformMatrix`
+    pub transform: Option<(&'a B::Buffer, Offset)>,
+}
+
+// VkTransformMatrixKHR
+#[derive(Debug)]
+pub struct TransformMatrix {
+    /// 3x4 row-major affine transformation matrix. Use `mint::RowMatrix3x4` if available.
+    pub matrix: [[f32; 4]; 3],
+}
+
+#[derive(Debug)]
+pub struct GeometryAabbs<'a, B: Backend> {
+    /// Must point to a buffer with buffer with an array of `AabbPositions`s
+    /// TODO: document lifetime required for this buffer
+    pub buffer: &'a B::Buffer,
+    pub buffer_offset: Offset,
+    pub buffer_stride: Offset,
+}
+
+// TODO doc
+#[derive(Debug)]
+pub struct AabbPositions {
+    /// Use `mint::Vector3` if available.
+    pub min: [f32; 3],
+    pub max: [f32; 3],
+}
+
+/// https://www.khronos.org/registry/vulkan/specs/1.2-extensions/man/html/VkAccelerationStructureGeometryInstancesDataKHR.html
+#[derive(Debug)]
+pub struct GeometryInstances<'a, B: Backend> {
+    /// Must point to a buffer with buffer with an array of `Instance`s
+    /// TODO: document lifetime required for this buffer
+    /// TODO this struct also allows passing an array of pointers, idk if that makes sense outside the host operations case
+    pub buffer: &'a B::Buffer,
+    pub buffer_offset: Offset,
+}
+
+bitflags! {
+    /// https://www.khronos.org/registry/vulkan/specs/1.2-extensions/man/html/VkGeometryInstanceFlagBitsKHR.html
+    pub struct InstanceFlags: u32 {
+        const TRIANGLE_FACING_CULL_DISABLE_BIT = 0x1;
+        const TRIANGLE_FRONT_COUNTERCLOCKWISE_BIT = 0x2;
+        const FORCE_OPAQUE_BIT = 0x4;
+        const FORCE_NO_OPAQUE_BIT = 0x8;
+    }
+}
+
+// TODO AFAIK rust doesn't have custom sized fields, so we'll need some binary writer wrapper to actually support this at an API level.
+#[derive(Debug)]
+pub struct Instance {
+    transform: TransformMatrix,
+    /// 24 bits
+    instance_custom_index: u32,
+    /// 8 bits visibility mask for the geometry. The instance may only be hit if rayMask & instance.mask != 0
+    mask: u32,
+    /// 24 bit
+    instance_shader_binding_table_record_offset: u32,
+    /// 8 bits
+    flags: InstanceFlags,
+
+    /// either B::AccelerationStructure (host operations) or GPU address (buffer + offset?)
+    acceleration_structure_reference: u64,
+}
+
+#[derive(Debug)]
+pub struct AccelerationStructureSizeRequirements {
+    pub acceleration_structure_size: u64,
+    pub update_scratch_size: u64,
+    pub build_scratch_size: u64,
+}
+
+#[derive(Debug)]
+pub enum AccelerationStructureCopyMode {
+    Clone,
```

Comment:
bikeshed: "clone" is the vk terminology, but I think in Rust it implies a deep copy, which I don't believe is what the acceleration structure copy command does.

Reply:
Indeed, let's rename. `Shallow`? `ShallowClone`?

Reply:
Any strong opinions against just `Copy`? I do think simpler is probably better, but I don't think even `Clone` would confuse many people familiar with gfx/vulkan...

Reply:
No strong opinion, `Copy` is fine :)